charms/worker/k8s/src/containerd.py

```diff
@@ -0,0 +1,388 @@
+        else:
+            ca_file_path.unlink(missing_ok=True)
+
+        cert_file_path = self.cert_file_path
+        if self.cert_file:
+            log.debug("Configure client certificate path %s", cert_file_path)
+            _ensure_file(cert_file_path, self.cert_file, 0o600, 0, 0)
+        else:
+            cert_file_path.unlink(missing_ok=True)
+
+        key_file_path = self.key_file_path
+        if self.key_file:
+            log.debug("Configure client key path %s", key_file_path)
+            _ensure_file(key_file_path, self.key_file, 0o600, 0, 0)
+        else:
+            key_file_path.unlink(missing_ok=True)
+
+    def ensure_hosts_toml(self):
+        """Ensure hosts.toml file."""
+        hosts_toml_path = self.hosts_toml_path
+        log.debug("Configure hosts.toml %s", hosts_toml_path)
+        _ensure_file(hosts_toml_path, tomli_w.dumps(self.hosts_toml), 0o600, 0, 0)
+
+
+class RegistryConfigs(pydantic.BaseModel, extra=pydantic.Extra.forbid):
+    """Represents a set of containerd registries.
+
+    Attrs:
+        registries (List[Registry]):
+    """
+
+    registries: List[Registry]
+
+
+def parse_registries(json_str: str) -> List[Registry]:
+    """Parse registry configurations from json string.
+
+    Args:
+        json_str (str): raw user supplied content
+
+    Returns:
+        RegistryConfigs parsed from json_str
+
+    Raises:
+        ValueError: if configuration is not valid
+    """
+    if not json_str:
+        return []
+
+    try:
+        parsed = json.loads(json_str)
+    except json.JSONDecodeError as e:
+        raise ValueError(f"not valid JSON: {e}") from e
+
+    parsed = RegistryConfigs(registries=parsed)
+    dupes = [x for x, y in collections.Counter(x.host for x in parsed.registries).items() if y > 1]
+    if len(dupes):
+        raise ValueError(f"duplicate host definitions: {','.join(dupes)}")
+    return parsed.registries
+
+
+def ensure_registry_configs(registries: List[Registry]):
+    """Ensure containerd configuration files match the specified registries.
+
+    Args:
+        registries (List[Registry]): list of registries
+    """
+    auth_config: Dict[str, Any] = {}
+    unneeded = {host.parent.name for host in CONFIG_PATH.glob("**/hosts.toml")}
+    for r in registries:
+        unneeded -= {r.host}
+        log.info("Configure registry %s (%s)", r.host, r.url)
+        r.ensure_certificates()
+        r.ensure_hosts_toml()
+
+        if r.username and r.password:
+            log.debug("Configure username and password for %s (%s)", r.url, r.host)
+            auth_config.update(r.auth_config)
+
+    for h in unneeded:
+        log.info("Removing unneeded registry %s", r)
+        (CONFIG_PATH / h / "hosts.toml").unlink(missing_ok=True)
+
+    if not auth_config and not unneeded:
+        return
+
+    registry_configs = {
+        "plugins": {"io.containerd.grpc.v1.cri": {"registry": {"configs": auth_config}}}
+    }
```

Comment:
these auths can be added in separate files under `/var/snap/k8s/common/etc/containerd/conf.d/*.toml`, so that you don't have to edit the main file, and thus avoid conflicts with other editors.

Reply:
ahh, thanks.  This was a port from microk8s and i missed that detail.

Reply:
After some hard-won debugging, i did manage to get user/password image pull validation via the Authentication Headers.  No tests yet included here though